protocol/README.md

```diff
@@ -0,0 +1,405 @@
+of MQTT are not supported. Message data is encoded using JSON within the
+MQTT messages.
+
+Each message is published to a defined MQTT topic and has a QoS value of
+`1`.
+
+Each message is encoded using UTF-8. The Payload Format Indicator must
+be set to `1` to indicate a UTF-8 encoded payload. The `contentType`
+must be set to `application/vnd.astoria+json`.
+
+Astoria services must set their MQTT client identifier to be the same
+value as the service *slug*.
+
+MQTT may be over TCP or a Websocket, although it is expected that
+services connect to the broker using MQTT over TCP.
+
+## Protocol Message Types
+
+There are three IPC mechanisms defined in the Astoria Protocol:
+
+-   State Updates
+-   Request-Response
+-   Streams
+
+### State Updates
+
+Services must publish their current state as soon as possible upon
+starting and then whenever the state is changed.
+
+The exact state that a service publishes varies per service and is
+defined in the protocol specification.
+
+Services must publish their state to the topic: `astoria/<slug>` where
+`<slug>` is the defined *slug* for that service.
+
+When a service is unavailable, it must publish an empty message to the
+state topic. An empty message on the state topic must also be set as the
+Last Will and Testament for the service's MQTT client.
+
+MQTT messages for state updates should be sent with the retained flag.
+
+The schema of a state update should be as follows:
+
+``` json
+{
+    "version": "1.2.3",
+    "protocol_version": "1.0.0",
+    "service": "example",
+    "type": "state_update",
+    "state": {}
+}
+```
+
+-   `protocol_version` - protocol version for the service
+-   `type` - always `state_update` for state updates.
+-   `service` - *slug* for the service
+-   `version` - software version for the service
+-   `state`- current state of the service, schema varies per service.
+
+### Request-Response
+
+Services may allow other services or clients to make a request against
+the service.
+
+Services must listen on the *request topic* `astoria/<slug>/request/+`
+where `<slug>` is the *slug* for the service. The wildcard at the end of
+the topic is for the *request slug*, a unique identifier for the
+request, e.g `kill_usercode`.
+
+A request can be made to a service by publishing a message to the
+appropriate *request topic*. The message may have the MQTT *response
+topic* set.
+
+The message content for a request must be a valid JSON object. The
+schema for the JSON object can vary depending on the request, and may be
+an empty object if no parameters are required.
+
+If the *response topic* is set on a request message, the service must
+respond by publishing a *response message* to the *response topic*. If
+no *response topic* is set, the service does not need to response, but
+should still attempt to perform the requested action.
+
+A *response topic* must be either of the format
+`astoria/<slug>/response/+` or be outside of the `astoria` topic
```

Comment:
> or be outside of the `astoria` topic

Suggestion: Let's not do this. Keeping things namespaced avoids any other interactions with outside systems. If we need to talk on another topic, the service can deal with that itself.

Reply:
[9d1bf66](https://github.com/srobo/astoria/pull/204/commits/9d1bf6690f1dde6b6b332bb8480c5680f91b2676)